clang/lib/CodeGen/CGStmt.cpp

```diff
@@ -1230,11 +1230,26 @@ CodeGenFunction::EmitCXXForRangeStmt(const CXXForRangeStmt &S,
   JumpDest LoopExit = getJumpDestInCurrentScope("for.end");
 
   LexicalScope ForScope(*this, S.getSourceRange());
+  const DeclStmt *RangeDS = cast<DeclStmt>(S.getRangeStmt());
+  const VarDecl *RangeVar = cast<VarDecl>(RangeDS->getSingleDecl());
+  if (getLangOpts().CPlusPlus23)
```

Comment:
I think we need to continue to improve here. We can decide whether to extend the lifetime in Sema. For example: do not store temporaries into ExpressionEvaluationContext::ForRangeLifetimeExtendTemps when building for-range-init expressions.

Reply:
What makes temporaries in for-range statements different from other temporaries?  The standard explicitly defines for-range in terms of a regular for-loop; it shouldn't require any special handling unless our AST is somehow deficient.

Reply:
Thanks for your review! [P2718R0](https://wg21.link/P2718R0) introduce lifetime extensions in for-range loops. Basic support for P2718R0 has been implemented in https://github.com/llvm/llvm-project/pull/76361 and everything works fine. When I works on this PR, I found that the lifetime of temporaries in`CXXDefaultInitExpr` was not extended, even if `MaterializedTemporaryExpr` sets ExtendingDecl and bound to `__range` variable.  So I thought about my current approach and I'm not sure if this is the correct approach. Can you provide a direction for improvement? Should we modify the AST instead of handling it in CodeGen?

Reply:
In `InitListChecker::FillInEmptyInitForField`, we will create a `CXXDefaultInitExpr` and fill in empty field `const A a& = A{};` for `for (auto &&x : B{}) {}` . Also there will have a `ExprWithCleanup` in `CXXDefaultInitExpr`. 
This is what I know about the difference in AST between the two cases, when I change the ExtendingDecl of `MaterializedTemporaryExpr`, it does not generate correct cleanups in CodeGen.

Reply:
The AST in the `B{A()}` case looks like the following.  If your patch changes the AST for `B{}` to look the same, it should be fine, I think?  From your description, I'm not exactly sure what's different about what CXXDefaultInitExpr rewriting generates.

```
      | `-VarDecl 0xc4927c0 <col:113, col:118> col:113 implicit used __range1 'B &&' cinit
      |   `-ExprWithCleanups 0xc492aa8 <col:113, col:118> 'B' xvalue
      |     `-MaterializeTemporaryExpr 0xc492a00 <col:113, col:118> 'B' xvalue extended by Var 0xc4927c0 '__range1' 'B &&'
      |       `-CXXFunctionalCastExpr 0xc492678 <col:113, col:118> 'B' functional cast to B <NoOp>
      |         `-InitListExpr 0xc492508 <col:114, col:118> 'B'
      |           `-MaterializeTemporaryExpr 0xc492568 <col:115, col:117> 'const A' lvalue extended by Var 0xc4927c0 '__range1' 'B &&'
      |             `-ImplicitCastExpr 0xc492550 <col:115, col:117> 'const A' <NoOp>
      |               `-CXXBindTemporaryExpr 0xc4718c8 <col:115, col:117> 'A' (CXXTemporary 0xc4718c8)
      |                 `-CXXTemporaryObjectExpr 0xc471890 <col:115, col:117> 'A' 'void ()'
```

Reply:
The AST of `B{}` looks like the following. As far as I know, when I try to debug the code generation of the temporary `Cleanups`, I found the following steps.

Let M1 be `MaterializeTemporaryExpr 0x133830b80 <col:19, col:22> 'B' xvalue extended by Var 0x133830940 '__range1' 'B &&'`
Let M2 be `MaterializeTemporaryExpr 0x1338305f0 <col:16, col:21> 'const A' lvalue extended by Var 0x133830940 '__range1' 'B &&'`
Let C1 be `ExprWithCleanups 0x133830be8 <col:19, col:22> 'B' xvalue`.
Let C2 be `ExprWithCleanups 0x133830658 <line:10:16, col:21> 'const A' lvalue`

for each lifetime extended `MaterializeTemporaryExpr`
1. Handling M2, `CGF.pushLifetimeExtendedDestroy ` push a EHCleanup into `EHStack` and push a NormalCleanup into `LifetimeExtendedCleanupStack`.
2. Handling C2, pop out the cleanup in `EHStack` and then push the copy of cleanup which in `LifetimeExtendedCleanupStack` into `EHStack`, and resize `LifetimeExtendedCleanupStack` to old size(0).
3. Handling M1, `CGF.pushLifetimeExtendedDestroy ` push a EHCleanup into `EHStack` and push a NormalCleanup into `LifetimeExtendedCleanupStack`, now `EHStack` has 2 cleanups(~B(), ~A()) and `LifetimeExtendedCleanupStack` has 1 cleanup(~B()).
4. Handling C1, pop out the top cleanup in `EHStack`, then emit the 2nd cleanup(emit call of ~A() in a `entry` BB), finally, push the copy of cleanup(~B()) which in `LifetimeExtendedCleanupStack` into `EHStack`, and resize `LifetimeExtendedCleanupStack` to old size(0).
5. Back to EmitCXXForRangeStmt, there are 1 cleanup(~B()) in `EHStack`, then `ForScope.ForceCleanup()` will emit this cleanup(call of ~B() in BB `for.cleanup`).

After the above steps, even if the lifetime of A is bound to `__range`, it is still not emit the expected cleanup.


```
`-CXXForRangeStmt 0x1338326d0 <line:17:3, col:36>
      |-<<<NULL>>>
      |-DeclStmt 0x133830d50 <col:19>
      | `-VarDecl 0x133830940 <col:19, col:22> col:19 implicit used __range1 'B &&' cinit
      |   `-ExprWithCleanups 0x133830be8 <col:19, col:22> 'B' xvalue
      |     `-MaterializeTemporaryExpr 0x133830b80 <col:19, col:22> 'B' xvalue extended by Var 0x133830940 '__range1' 'B &&'
      |       `-CXXFunctionalCastExpr 0x1338307e0 <col:19, col:22> 'B' functional cast to B <NoOp>
      |         `-InitListExpr 0x133830408 <col:20, col:22> 'B'
      |           |-IntegerLiteral 0x133829da0 <col:21> 'int' 0
      |           `-CXXDefaultInitExpr 0x133830670 <col:22> 'const A' lvalue has rewritten init
      |             `-ExprWithCleanups 0x133830658 <line:10:16, col:21> 'const A' lvalue
      |               `-MaterializeTemporaryExpr 0x1338305f0 <col:16, col:21> 'const A' lvalue extended by Var 0x133830940 '__range1' 'B &&'
      |                 `-ImplicitCastExpr 0x1338305d8 <col:16, col:21> 'const A' <NoOp>
      |                   `-CXXFunctionalCastExpr 0x1338305b0 <col:16, col:21> 'A' functional cast to A <NoOp>
      |                     `-CXXBindTemporaryExpr 0x133830590 <col:17, col:21> 'A' (CXXTemporary 0x133830590)
      |                       `-InitListExpr 0x1338304f8 <col:17, col:21> 'A'
      |                         `-InitListExpr 0x133830540 <col:18, col:20> 'int[1]'
      |                           `-IntegerLiteral 0x1338297e8 <col:19> 'int' 0
```

Reply:
The problem seems to be caused by C2 `ExprWithCleanups `

Reply:
It makes sense that the extra ExprWithCleanups would force the destructor to run early, since that's supposed to be the marker for a full-expression.

Do we actually need to handle this case?  Looking at the current standard, it says "A temporary expression bound to a reference member from a default member initializer is ill-formed", which would seem to imply we should reject any case that would require this sort of lifetime extension.

Reply:
I've another separate PR https://github.com/llvm/llvm-project/pull/87933 that support lifetime extension of temporary created by aggregate initialization using a default member initializer. In this case the temporary `A{}` created by aggregate initialization(I've misunderstand here before), I think since DR1815, it's well-formed.

Reply:
If you think "A temporary expression bound to a reference member from a default member initializer is ill-formed" doesn't refer to this case, what does it refer to?

Reply:
Thr quoted rule is [class.base.init]/11, and in that context is describing the behavior of the initialization of class members in a constructor.

Aggregate initialization is handled by the general rule in [class.temporary] but see also the [note in [dcl.init.general]](https://eel.is/c++draft/dcl.init#general-16.6.2.2.sentence-8) that explicitly mentions this case.

Reply:
Thank you for your confirmation, I will add a special-case to ExprWithCleanups handling once PR  https://github.com/llvm/llvm-project/pull/87930 https://github.com/llvm/llvm-project/pull/87933 merged.

Reply:
> The AST in the `B{A()}` case looks like the following. If your patch changes the AST for `B{}` to look the same, it should be fine, I think? From your description, I'm not exactly sure what's different about what CXXDefaultInitExpr rewriting generates.
> 
> ```
>       | `-VarDecl 0xc4927c0 <col:113, col:118> col:113 implicit used __range1 'B &&' cinit
>       |   `-ExprWithCleanups 0xc492aa8 <col:113, col:118> 'B' xvalue
>       |     `-MaterializeTemporaryExpr 0xc492a00 <col:113, col:118> 'B' xvalue extended by Var 0xc4927c0 '__range1' 'B &&'
>       |       `-CXXFunctionalCastExpr 0xc492678 <col:113, col:118> 'B' functional cast to B <NoOp>
>       |         `-InitListExpr 0xc492508 <col:114, col:118> 'B'
>       |           `-MaterializeTemporaryExpr 0xc492568 <col:115, col:117> 'const A' lvalue extended by Var 0xc4927c0 '__range1' 'B &&'
>       |             `-ImplicitCastExpr 0xc492550 <col:115, col:117> 'const A' <NoOp>
>       |               `-CXXBindTemporaryExpr 0xc4718c8 <col:115, col:117> 'A' (CXXTemporary 0xc4718c8)
>       |                 `-CXXTemporaryObjectExpr 0xc471890 <col:115, col:117> 'A' 'void ()'
> ```

I have addressed this suggestion, seems it's works fine now, and avoid the changes in CodeGen.